File: codes/kotlin/xxx_chaos.kt
@Suppress("UNUSED_VARIABLE", "UNCHECKED_CAST", "USELESS_CAST")

fun main() {
    fun jsonStringify(anything: Any? = null, pretty: Boolean = false, indent: String = "    "): String {
        var indentLevel = 0
        fun jsonStringifyInner(anythingInner: Any?, indentInner: String): String {
            if (anythingInner == null) return "null"
            if (anythingInner is String) return "\"${anythingInner}\""
            if (anythingInner is Number || anythingInner is Boolean) return "${anythingInner}"
            if (anythingInner is MutableList<*>) {
                if (anythingInner.size == 0) return "[]"
                indentLevel += 1
                var result = (if (pretty == true) "[\n${indentInner.repeat(indentLevel)}" else "[")
                for ((arrayItemIndex, arrayItem) in anythingInner.withIndex()) {
                    result += jsonStringifyInner(arrayItem, indentInner)
                    if ((arrayItemIndex + 1) != anythingInner.size) result += (if (pretty == true) ",\n${indentInner.repeat(indentLevel)}" else ", ")
                }
                indentLevel -= 1
                result += (if (pretty == true) "\n${indentInner.repeat(indentLevel)}]" else "]")
                return result
            }
            if (anythingInner is MutableMap<*, *>) {
                if (anythingInner.entries.size == 0) return "{}"
                indentLevel += 1
                var result = (if (pretty == true) "{\n${indentInner.repeat(indentLevel)}" else "{")
                anythingInner.entries.forEachIndexed { objectEntryIndex, objectEntry ->
                    val objectKey = objectEntry.key
                    val objectValue = objectEntry.value
                    result += "\"${objectKey}\": ${jsonStringifyInner(objectValue, indentInner)}"
                    if ((objectEntryIndex + 1) != anythingInner.entries.size) result += (if (pretty == true) ",\n${indentInner.repeat(indentLevel)}" else ", ")
                }
                indentLevel -= 1
                result += (if (pretty == true) "\n${indentInner.repeat(indentLevel)}}" else "}")
                return result
            }
            return "null"
        }
        return jsonStringifyInner(anything, indent)
    }

    fun arrayReduce(callbackFunction: (Any?, Any?, Int, MutableList<Any?>) -> Any?, anArray: MutableList<Any?>, initialValue: Any?): Any? {
        // JavaScript-like Array.reduce() function
        var result = initialValue
        for ((arrayItemIndex, arrayItem) in anArray.withIndex()) {
            result = callbackFunction(result, arrayItem, arrayItemIndex, anArray)
        }
        return result
    }

    println("\n// JavaScript-like Array.reduce() in Kotlin MutableList")
    
    val numbers = mutableListOf<Any?>(36, 57, 2.7, 2.3, -12, -34, -6.5, -4.3)
    println("numbers: ${jsonStringify(numbers)}")

    var numbersTotal: Any?

    println("// using JavaScript-like Array.reduce() function \"arrayReduce\"")

    numbersTotal = arrayReduce({ currentResult: Any?, currentNumber: Any?, _: Int, _: MutableList<Any?> -> 
        when {
            currentResult is Number && currentNumber is Number -> (currentResult.toDouble() + currentNumber.toDouble())
            else -> currentResult
        }
    }, numbers, 0.0)
    println("total number: ${numbersTotal}")
    // total number: 41.2
}

File: codes/swift/xxx_chaos.swift
import Foundation

typealias MyObject = [String: Any?]
typealias MyArray = [Any?]

func arrayReduce(_ callbackFunction: (Any?, Any?, Int, MyArray) -> Any?, _ anArray: MyArray, _ initialValue: Any?) -> Any? {
    // JavaScript-like Array.reduce() function
    var result = initialValue
    for (arrayItemIndex, arrayItem) in anArray.enumerated() {
        result = callbackFunction(result, arrayItem, arrayItemIndex, anArray)
    }
    return result
}

func getDouble(_ anything: Any?) -> Double? {
    guard let anything = anything else {
        return nil
    }
    if let doubleValue = anything as? Double {
        return doubleValue
    }
    if let floatValue = anything as? Float {
        return Double(floatValue)
    }
    if let floatValue = anything as? Float16 {
        return Double(floatValue)
    }
    if let floatValue = anything as? Float32 {
        return Double(floatValue)
    }
    if let floatValue = anything as? Float64 {
        return Double(floatValue)
    }
    if let floatValue = anything as? Float80 {
        return Double(floatValue)
    }
    if let intValue = anything as? Int {
        return Double(intValue)
    }
    if let intValue = anything as? Int8 {
        return Double(intValue)
    }
    if let intValue = anything as? Int16 {
        return Double(intValue)
    }
    if let intValue = anything as? Int32 {
        return Double(intValue)
    }
    if let intValue = anything as? Int64 {
        return Double(intValue)
    }
    if let uintValue = anything as? UInt {
        return Double(uintValue)
    }
    if let uintValue = anything as? UInt8 {
        return Double(uintValue)
    }
    if let uintValue = anything as? UInt16 {
        return Double(uintValue)
    }
    if let uintValue = anything as? UInt32 {
        return Double(uintValue)
    }
    if let uintValue = anything as? UInt64 {
        return Double(uintValue)
    }
    return nil
}

let numbers: MyArray = [36, 57, 2.7, 2.3, -12, -34, -6.5, -4.3]

var numbersTotal: Any?

numbersTotal = arrayReduce({ (currentResult: Any?, currentNumber: Any?, _: Int, _: MyArray) -> Any? in
    if let currentResultDouble = getDouble(currentResult), let currentNumberDouble = getDouble(currentNumber) {
        return currentResultDouble + currentNumberDouble
    }
    return currentResult
}, numbers, 0.0)
print("total number: \(numbersTotal ?? "nil")")
// total number: 41.2

numbersTotal = numbers.reduce(0.0) { (currentResult: Any?, currentNumber: Any?) -> Any? in
    if let currentResultDouble = getDouble(currentResult), let currentNumberDouble = getDouble(currentNumber) {
        return currentResultDouble + currentNumberDouble
    }
    return currentResult
}
print("total number: \(numbersTotal ?? "nil")")
// total number: 41.2
